js/privatebin.js

```diff
@@ -3666,6 +3667,30 @@ jQuery.PrivateBin = (function($, RawDeflate) {
             newDoc.close();
         }
 
+        /**
+         * download text
+         *
+         * @name   TopNav.downloadText
+         * @private
+         * @function
+         */
+        function downloadText()
+        {
+	    var filename='paste.txt';
```

Comment:
Okay, let's think about the use cases.

IMHO, the main one I see is:
* sharing source code snippets and getting them as a file
* sharing any other text and getting it as a file

As such, the last one is clearly fulfilled. However, another thing:
* When I as a user downloaded multiple pastes, I'd like

So proposals:
* Could we include some identification thing in the filename? E.g. the paste ID (I know it's ugly, but it's better than `paste.txt`). Or maybe extract the first line up to… 20(?)… characters and put it into it.
  The provided [example](https://paste.chapril.org/?ef6a1b876b5c4526#4a9cGPmXJMzZyPW4diMpvesvx3x8x1NqCuidn6raP8Kn) would get `paste-ef6a1b876b5c4526.txt` e.g. (you could at least find it in your browser history then) or `paste-       /__.txt` – oh, I see we would have to rmeove special chars and likely trim spaces, so uhm… yeah… hard.
  The second approach would be more useful if you have a Markdown paste, though, which may start with a nice header in Markdown `paste-the_key_to_everything.txt` e.g.
  * If it is set to source code mode, could we auto-detect the file extension? Does our prettifer library maybe provide a way to do this? Having `paste.js` is clearly way nicer, than `paste.txt`.
  
I guess the best implementation (UX wise etc.) would take the paste type into account and:
* for code: `paste.[appropiate file extension | fallback to txt]`
* for Markdown: `paste-[first x chars of paste, sanitized to remove special chars for all OSes].md`
* for text only: `paste.txt`

Reply:
> Or maybe extract the first line up to… 20(?)… 
Bad idea, you can't know what in that text file can be, so here can be things with slashes.

Reply:
Well… if we limit ourselve to latin alphabets (i.e. `\w` or so) that is not that hard to do: https://stackoverflow.com/a/8485137/5008962
But yeah, I dunno…

Reply:
Hah, and in result someone will get weird filename like "paste-I've stolen your files. txt"? :rofl:

Reply:
Thank you for these suggestions. Depending on the browser/settings, the user will be asked where to save his file to, at which point they can change the suggested filename.

With this in mind, I agree that it would make sense to add the ID, as it is unique to the paste. This would cover these scenarios:
- browser downloads automatically to default location -> the filename is unique and the user can figure out which paste URL it belongs to if they download multiple pastes (that would currently result in files called, i.e. paste.txt, paste (2).txt, paste (3).txt, etc.)
- browser asks user for location to store file -> user has a unique per paste filename preset, can optionally change it or just confirm name with the "Save" button
- If the same paste gets downloaded again (by accident?):
  - browser downloads automatically to default location -> the filename gets a suffix appended - not ideal, but at least they still know which paste they downloaded twice
  - browser asks user for location to store file -> user gets an error message about the file already existing and can then either abort the download or rename the file

Reply:
Applied the ID suggestion in bc114522592f0c3ed2d366363ad264437cf904aa - let me know if you'd like further changes for this point.

Reply:
> Depending on the browser/settings, the user will be asked where to save his file to, at which point they can change the suggested filename.

No browser does this AFAIK unless no download location is set (but it is set by default…).

I'm fine with the ID though and it is the easiest thing to implement, so okay…